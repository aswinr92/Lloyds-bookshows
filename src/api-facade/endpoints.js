export const moviesEndpoint =
  "https://www.omdbapi.com/?apikey=4b6c78de&s=of%20the&y=2020&type=movie";

export const detailsEndpoint = "https://omdbapi.com/?apikey=4b6c78de&i=";
