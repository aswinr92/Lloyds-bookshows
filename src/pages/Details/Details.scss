@import "../../colors.scss";
.details {
  display: flex;
  flex-direction: column;
  background: white;
  border-radius: 20px;
  margin: 20px 40px;
  color: $grey;
  padding: 20px;
  .meta {
    border-bottom: 1px solid #dcdcdc;
    margin-bottom: 5px;
    padding-bottom: 5px;
    &__right {
      margin-left: 20px;
    }
    &__title {
      font-size: 20px;
      color: $grey;
      margin-bottom: 10px;
    }
    &__content {
      display: flex;
      flex-wrap: wrap;
    }
    &__details {
      font-size: 14px;
    }
    &__desc {
      color: $grey;
      font-size: 15px;
      margin-bottom: 20px;
    }
  }
  .theatres {
    padding: 20px;
    &__label {
      font-weight: bold;
      font-size: 16px;
      margin-bottom: 5px;
    }
  }
  .actions {
    margin-top: 10px;
  }
}
