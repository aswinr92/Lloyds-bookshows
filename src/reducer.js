function reducer(state = { movie: "", theatre: "", time: "" }, action) {
  switch (action.type) {
    case "SELECT_MOVIE":
      return {
        ...state,
        movie: action.payload
      };
    case "SELECT_THEATRE":
      return {
        ...state,
        theatre: action.payload
      };
    case "SELECT_TIME":
      return {
        ...state,
        time: action.payload
      };
    default:
      return state;
  }
}

export default reducer;
