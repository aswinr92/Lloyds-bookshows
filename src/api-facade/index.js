import { moviesEndpoint, detailsEndpoint } from "./endpoints";

export const getMovies = async () => {
  const response = await fetch(moviesEndpoint);
  return await response.json();
};

export const getMovieDetails = async id => {
  const response = await fetch(`${detailsEndpoint}${id}`);
  return await response.json();
};
