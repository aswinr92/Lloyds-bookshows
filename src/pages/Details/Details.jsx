import React from "react";
import { Link, useParams } from "react-router-dom";
import { getMovieDetails } from "../../api-facade";
import theatres from "../../data/theatres.json";
import "./Details.scss";
import Accordion from "../../components/Accordion";

const Details = () => {
  let { id } = useParams();

  const [details, setDetails] = React.useState();
  const [selected, setSelected] = React.useState({ theatre: "", timing: "" });

  const getDetails = async () => {
    const response = await getMovieDetails(id);
    setDetails(response);
  };

  const handleOpen = t => {
    setSelected({ theatre: t });
  };

  const handleSelect = t => {
    const s = selected.timing === t ? false : t;
    setSelected({ ...selected, timing: s });
  };

  React.useEffect(() => {
    getDetails();
  });

  return (
    <div className="details">
      {details && (
        <div>
          <div className="meta">
            <div className="meta__title">{details.Title}</div>
            <div className="meta__content">
              <img
                src={details.Poster}
                alt={details.Title}
                height="250"
                width="250"
              />
              <div className="meta__right">
                <div className="meta__desc">{details.Plot}</div>
                <div className="meta__details">
                  <div className="meta__ratings">
                    Rating: {details.Ratings[0].Value}
                  </div>
                  <div className="meta__released">
                    Released On: {details.Released}
                  </div>
                  <div className="meta__rated">Rated: {details.Rated}</div>
                  <div className="meta__runtime">Runtime:{details.Runtime}</div>
                </div>
              </div>
            </div>
          </div>

          <div className="theatres">
            <div className="theatres__label">Now Playing in...</div>

            {theatres.map(t => (
              <Accordion
                {...t}
                handleOpen={handleOpen}
                open={t.name === selected.theatre}
                handleSelect={handleSelect}
              />
            ))}
            <div className="actions">
              <Link
                className={`select__button cta__button ${
                  selected.theatre && selected.timing ? "active" : ""
                }`}
                to={`/theatre/${selected.theatre}/${id}/${selected.timing}`}
              >
                Select Seats
              </Link>
            </div>
          </div>
        </div>
      )}
    </div>
  );
};

export default Details;
