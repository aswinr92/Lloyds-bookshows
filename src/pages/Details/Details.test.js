import { mount } from "enzyme";
import * as React from "react";
import { HashRouter as MockRouter } from "react-router-dom";
import Details from "./Details";
import fetchMock from "fetch-mock";
import { detailsEndpoint } from "../../api-facade/endpoints";
import { Provider } from "react-redux";
import { createStore } from "redux";
import reducer from "../../reducer";

describe("Details Page", () => {
  afterEach(fetchMock.restore);
  const a = {
    Title:
      "Birds of Prey: And the Fantabulous Emancipation of One Harley Quinn",
    Year: "2020",
    Rated: "R",
    Released: "07 Feb 2020",
    Runtime: "109 min",
    Genre: "Action, Adventure, Crime",
    Director: "Cathy Yan",
    Writer: "Christina Hodson",
    Actors:
      "Margot Robbie, Rosie Perez, Mary Elizabeth Winstead, Jurnee Smollett-Bell",
    Plot:
      "After splitting with the Joker, Harley Quinn joins superheroes Black Canary, Huntress and Renee Montoya to save a young girl from an evil crime lord.",
    Language: "English, Chinese",
    Country: "USA, UK",
    Awards: "3 nominations.",
    Poster:
      "https://m.media-amazon.com/images/M/MV5BMzQ3NTQxMjItODBjYi00YzUzLWE1NzQtZTBlY2Y2NjZlNzkyXkEyXkFqcGdeQXVyMTkxNjUyNQ@@._V1_SX300.jpg",
    Ratings: [
      { Source: "Internet Movie Database", Value: "6.6/10" },
      { Source: "Rotten Tomatoes", Value: "79%" },
      { Source: "Metacritic", Value: "60/100" }
    ],
    Metascore: "60",
    imdbRating: "6.6",
    imdbVotes: "42,728",
    imdbID: "tt7713068",
    Type: "movie",
    DVD: "N/A",
    BoxOffice: "N/A",
    Production: "Warner Bros. Pictures",
    Website: "N/A",
    Response: "True"
  };
  const store = createStore(reducer, {
    movie: "1",
    theatre: "",
    time: ""
  });
  it("should render the component", () => {
    const component = mount(
      <Provider store={store}>
        <MockRouter>
          <Details />
        </MockRouter>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });
  it("should fetch movie details and display title on load", async () => {
    fetchMock.get(`${detailsEndpoint}1`, a);

    const component = mount(
      <Provider store={store}>
        <MockRouter>
          <Details />
        </MockRouter>
      </Provider>
    );
    await delay();
    component.update();
    fetchMock.done();
    expect(component.find("div.meta__title").text()).toEqual(a.Title);
  });

  it("should let you book movie", async () => {
    fetchMock.get(`${detailsEndpoint}1`, a);

    const component = mount(
      <Provider store={store}>
        <MockRouter>
          <Details />
        </MockRouter>
      </Provider>
    );
    await delay();
    component.update();
    fetchMock.done();
    expect(component.find("div.select__button").hasClass("active")).toBe(false);
    component
      .find("div.accordion")
      .at(0)
      .find("div.icon")
      .simulate("click");
    await delay();
    component.update();
    component
      .find("div.time__button")
      .at(0)
      .simulate("click");
    expect(component.find("a.select__button").hasClass("active")).toBe(true);
  });
});

export function delay(milliseconds = 10) {
  return new Promise(resolve => setTimeout(resolve, milliseconds));
}
