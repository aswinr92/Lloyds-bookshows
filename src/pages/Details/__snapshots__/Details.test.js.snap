// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details Page should render the component 1`] = `ReactWrapper {}`;
